### PR TITLE
[RFC] Arrow Functions params have parens unless they implicitly return

### PR DESCRIPTION
**Pull request by rattrayalex**
As discussed in https://github.com/prettier/prettier/issues/812#issuecomment-313829292

I'm not sure this is the right move, and I think several of the changes may be regressions.

### CODE REVIEW

File: tests/arrows/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -326,11 +326,11 @@ const middleware = options => (req, res, next) => {
   // ...
 };
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-const fn = b => c => d => {
+const fn = b => c => (d) => {
```
Comment (rattrayalex) [inline tests/arrows/__snapshots__/jsfmt.spec.js.snap]: this doesn't seem great in the currying case, but special-casing would certainly be awful

File: tests/test_declarations/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -93,22 +93,26 @@ it(\`handles
   console.log("hello!");
 });
 
-test("does something really long and complicated so I have to write a very long name for the test", done => {
+test("does something really long and complicated so I have to write a very long name for the test", (
+  done
```
Comment (rattrayalex) [inline tests/test_declarations/__snapshots__/jsfmt.spec.js.snap]: this seems like a clear regression, not sure what the behavior is meant to be testing... probably worth looking into, as I imagine the intent is for this to all stay on one line.

Comment (azz) [inline tests/test_declarations/__snapshots__/jsfmt.spec.js.snap]: It is testing that the string literal doesn't break to the next line. There are special cases for `test`/`it`, etc.

Comment (azz) [inline tests/test_declarations/__snapshots__/jsfmt.spec.js.snap]: To fix this, `(`, `done`, and `)` should be in the same `group`.

File: tests/method-chain/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -61,7 +61,7 @@ it('should group messages with same created time', () => {
   });
 });
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-export default store => {
+export default (store) => {
```
Comment (azz) [inline tests/method-chain/__snapshots__/jsfmt.spec.js.snap]: This is one of few cases where I 100% prefer using parens.

File: tests/test_declarations/__snapshots__/jsfmt.spec.js.snap (hunk 2 above)
Comment (rattrayalex) [inline tests/test_declarations/__snapshots__/jsfmt.spec.js.snap]: hmm, k, I'll need to fix this.

**Comment by rattrayalex** [comment]
In #2473 @jlongster posted that he'd like to maintain the current behavior, so closing

**Pull request closed without merging**